### ISSUE TITLE
Return custom errors for most common cases instead of generic `Error` or `ValidationError`

### ISSUE THREAD

**Issue by ste93cry** [lestrrat-go/jwx#706]
**Abstract**
While building my own HTTP handler to verify a JWT token, I realized that besides some validation errors this lib always returns generic errors. The problem with this approach is that I have no way to know what the error really is, besides parsing the message, and this make things complex in case a specific error should be returned upstream. For example, I may want to return a different HTTP response if the parsing of the token has failed or if the signature is invalid.

**Describe the proposed solution/change**
Below I prepared a comparison between 3 common libs I found in the Internet to work with JWT. I listed the errors I would like to be able to distinguish while parsing a token and whether each lib returns a specific error type for each case or not.

| | `lestrrat-go/jwx` | `golang-jwt/jwt` | `square/go-jose` |
|-|-|-|-|
| `ErrTokenMalformed` | No | Yes | No |
| `ErrTokenSignatureInvalid` | No | Yes | No |
| `ErrTokenInvalidAudience` | Yes (no specific error) | Yes | Yes |
| `ErrTokenExpired` | Yes | Yes| Yes |
| `ErrTokenUsedBeforeIssued` | Yes | Yes | Yes |
| `ErrTokenInvalidIssuer` | Yes (no specific error) | Yes | Yes |
| `ErrTokenNotValidYet` | Yes | Yes | Yes |
| `ErrTokenInvalidId` | Yes (no specific error) | Yes | Yes |
| `ErrTokenInvalidClaims` | Yes (no specific error) | Yes | Yes |

From what you can see above, most of the errors are already handled, but it is impossible to distinguish them because they are all instances of `ValidationError`. The first two are really important to me, as I suspect that many users would like to return to the caller whether a token has an invalid format, and right now the error is just a plain `fmt.Errorf()`. Implementing custom errors for those cases would allow people to do:

```go
token, err := jwt.ParseString(...)

if err == nil {
    fmt.Println("Have a nice day")
} else if errors.Is(err, ErrTokenMalformed) {
    fmt.Println("That's not even a token")
} else if errors.Is(err, ErrTokenSignatureInvalid) {
    fmt.Println("You are trying to fool me")
} else {
    fmt.Println("Could not handle this token", err)
}
```

**Analysis**
A workaround may be to parse the error message and return a custom error depending on the outcome. This is the solution chosen by [`go-chi/jwtauth`](https://github.com/go-chi/jwtauth/blob/137d669fb7769f7c7b470957d4d8447e3e5989db/jwtauth.go#L144). Unfortunately, as you can imagine, it can easily break and it isn't maintenable in the long term.

**Comment by lestrrat** [comment]
I can see some missing error types from your table, but you can already do `errors.Is()`:

https://github.com/lestrrat-go/jwx/blob/7a86ab3acef7846016d103708bd327a8451f37b2/jwt/validate_test.go#L223

Or am I missing something?

**Comment by lestrrat** [comment]
It's also in the docs https://github.com/lestrrat-go/jwx/blob/develop/v2/docs/01-jwt.md#detecting-error-types

**Comment by ste93cry** [comment]
>I can see some missing error types from your table

You're right, while compiling the table I got confused: as you said some errors like `ErrTokenExpired` are already available. The rest though are treated just as plain `Error` or `ValidationError` errors, and if I'm not wrong there's no way to distinguish them besides their error message

**Comment by lestrrat** [comment]
First off, I don't agree that TokenMalformed and TokenSignatureInvalid should be specific errors. Those are parsing problems, and there are a myriad of ways that parsing can fail, and I don't agree that only these two should receive special treatment, nor that each minor error type be represented as an its own type.

Not sure about InvalidId (actually, I don't know that this means) and InvalidClaims either, but not necessarily against it either.

As for the rest, sure, but I'm not terribly excited about the idea so my motivation to implement this is meh: so PRs welcome

**Comment by ste93cry** [comment]
>I don't agree that TokenMalformed and TokenSignatureInvalid should be specific errors. Those are parsing problems, and there are a myriad of ways that parsing can fail, and I don't agree that only these two should receive special treatmentnor that each minor error type be represented as an its own type.

`TokenMalformed` should be thrown if parsing fails, for example if the number of segments of the JWT is not the required one or if a segment is not valid JSON. I don't expect it to be used to report specific situations besides the general "parsing error". Maybe for the end user it makes sense to return a `401 Unauthorized` in all cases, maybe a `400 Bad Request` would better fit to describe a problem with the format of the token string: if I don't know what happened though, it's really hard for me to do it.

`TokenSignatureInvalid` instead makes sense for me because if the signature does not pass verification, I'm stuck with a general error that tells me nothing about why the token verification failed: was it because the value was invalid or because the token was tampered?

>Not sure about InvalidId (actually, I don't know that this means) and InvalidClaims either

To be honest, I'm not interested in `ErrTokenInvalidId` either, but I listed it for completeness. This error should be related to the validation of standard [`jti`](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.7) claim, which now falls under the standard `ValidationError` error. Probably, we can skip implementing it until there's a real need for it. About `ErrTokenInvalidClaims`, there is now no way to know if : they all get treated as validation errors

**Comment by lestrrat** [comment]
Ah, is SignatureInvalid error for JWS verification failure? that's a possibility, then. But the name should definitely not be "SignatulreInvalid": I can take `VerificationError` or similar which contains the word Verify/Verfication.

Not convinced about TokenMalformed. Sounds like a verification error, validation error, and everything else should do the job.

**Comment by ste93cry** [comment]
>Ah, is SignatureInvalid error for JWS verification failure?

Yes, it is. About the name, it comes from `golang-jwt/jwt`, but I have no strong opinion, so if you think that `VerificationError` better fits this lib, then I'm all in!

>Not convinced about TokenMalformed

I'm not a native English speaker, so at my eyes the "malformed" word looks appealing. Otherwise, what about `ErrTokenFormatInvalid`?

**Comment by lestrrat** [comment]
No, it's not about the naming. I don't want to create specific error types for parsing failures.

This library consists of `jws`, `jwe`, `jwk`, `jwt`, and it tries very hard to be symmetric (i.e. $pkg.Parse for each package _feels_ uniform/consistent with respect to other packages). When you start creating specific error types for parsing errors, it opens up a can of worms to align other packages to do the same.

And this has two problems: (1) it's a lot of work and a lot of changes, and (2) I _generally_ don't like specific error types. JWT is a bit of an exception, but this is as far as I want to go. I am _not_ creating a specific error type for parsing errors.

You can already achieve what you want without a parsing error type:

```go
if err := jwt.Parse(...); err != nil {
   switch {
   case errors.Is(err, ....):
   case errors.Is(err, ....):
   default:
      fmt.Println("parse error!")
   }
}
```

**Comment by ste93cry** [comment]
I understand that symmetry is important,but in the end all those standards at some point must return an error if parsing fails, and the point is to return something that has a specific type so that I can intercept that kind of failure. If the name of the error is ErrParseFailure or something else, that is just a matter of agreement so that it's not too specific and not too general. Said this, if you are a strong opponent of this use case it's fine, but from what I understand though, you are open to changes for the error that occurs when verification fails, right?

>You can already achieve what you want without a parsing error type:

Right, but in that case I need to code a `switch` full of `case` statements that maybe I'm not even interested in just for the sake of excluding those error types. And what about the future, if another error will be introduced? The `default` statement will catch it as well at that point, when it should not

**Comment by lestrrat** [comment]
Sorry, the answer is no.

If you have more to say, please use (pseudo)code that represents exactly what you want to do / what you can't currently do.

**Comment by ste93cry** [comment]
Let's suppose that you are writing some tests to check that the function that parses and verify the JWT token works as expected. Let's also suppose that the test is a table-driven test, where the input is the token and the output is the expected error, if any. The final purpose of the the test is to ensure that the verification process takes into account the configuration passed via `jwt.With...()` options.

```go
func VerifyToken(token string) error {
    if _, err := jwt.ParseString(tc.token, jwt.WithKeySet(...), jwt.WithRequiredClaim(...)); err != nil {
        return err
    }

    return nil
}

func TestVerifyToken(testing *T) {
    tcs := []struct {
        token string
        wantErr error
    }{
        {
            token string
            wantErr: ???,
        },
    }

    for _, tc := range tcs {
        tc := tc

        t.Run("...", func(t *testing.T) {
            if _, err := VerifyToken(tc.token); err != nil {
                assert.ErrorIs(t, tc.wantErr, err)
            }
       }
    }
}
```

I can assign the custom errors to the `wantErr` field, e.g. `jwt.ErrTokenExpired()` and the assertion would work fine. However, as soon as I want to use `ValidationError`, there's no way to assign it to the field because of it not being an immutable error. Yes, I can change the way I do the assertion to use an anonymous function that calls `jwt.IsValidationError(`) or whatever else I need, but that seems just a lot more complicated than what it should be in my opinion. For example, I find to be a lot valuable to be able to write a test case to ensure that the required claims are taken into account when validating the token. Anyway, feel free to close this issue if you still don't see the value in having a custom error

**Comment by lestrrat** [comment]
> Anyway, feel free to close this issue if you still don't see the value in having a custom error

I wrote `As for the rest, sure, but I'm not terribly excited about the idea so my motivation to implement this is meh: so PRs welcome` so I'm not saying you can't have custom errors.

I'm saying I won't allow for parse (invalid format) errors, i.e. equivalent of TokenMalformed errors. You can send me a PR for your RequiredClaim error type, and I will happily accept it.

**Comment by ste93cry** [comment]
While trying to look at how to implement the `ErrRequiredClaimMissing` (or whatever we want to call it) I realized that the error message is dynamic because it contains the claim's name. However, this means that I cannot do something along the lines of `var ErrRequiredClaimMissing = NewValidationError(...)`. The other way of creating an error that I know is to implement a custom `struct`, however that would make `errors.Is()` impossible to be used. What do you suggest?

**Comment by lestrrat** [comment]
>however that would make errors.Is() impossible to be used

It's definitely possible. This is straight out of the `errors.Is` documentation:

>    An error type might provide an Is method so it can be treated as equivalent
>    to an existing error. For example, if MyError defines
>
>        func (m MyError) Is(target error) bool { return target == fs.ErrExist }
>
>    then Is(MyError{}, fs.ErrExist) returns true. See syscall.Errno.Is for an
>    example in the standard library.

Example whipped out in 5 min (i.e. not thoroughly tested, but I think it handles all cases):

https://go.dev/play/p/jfGA5DWYt58

**Comment by ste93cry** [comment]
Oh, completely missed out that little bit of the doc. Thanks for pointing it out, as a newbie of Go it will help me in the future as well

**Comment by ste93cry** [comment]
>You can send me a PR for your RequiredClaim error type, and I will happily accept it.

Opened it! Just wondering if you would be open to also add the errors `ErrTokenInvalidAudience`, `ErrTokenInvalidIssuer` and `ErrTokenInvalidId`: since there are already errors specific for the other common standard claims, for consistency it could be valuable to have them

**Comment by lestrrat** [comment]
> if you would be open to also add the errors...

Yes, but please follow the existing pattern (see #714). I _really_ don't want to expose error types.
This is because once you expose error types, it becomes really hard to modify them in the future. We had limited access to error types by only exposing an interface and functions to return the error for this. Things would be different if users could write `jwt.SomeRandomError{}` in their code.

**Comment by lestrrat** [comment]
closed by #714
